File: C++/serialize.cpp
#include <iostream>
#include <string>
#include <vector>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    std::string Serialize(TreeNode* root) {
        std::string serialized;
        serialize(root, serialized);
        return serialized;
    }

    TreeNode* Deserialize(const std::string& str) {
        int index = 0;
        return deserialize(str, index);
    }

private:
    void serialize(TreeNode* node, std::string& result) {
        if (!node) {
            result += "#,";
            return;
        }
        result += std::to_string(node->val) + ",";
        serialize(node->left, result);
        serialize(node->right, result);
    }

    TreeNode* deserialize(const std::string& str, int& index) {
        if (index >= str.length() || str[index] == ',' || str[index] == '#') {
            index++;
            return nullptr;
        }
        TreeNode* node = new TreeNode(std::stoi(str.substr(index, str.find(',', index) - index)));
        index = str.find(',', index) + 1; 
        node->left = deserialize(str, index);
        node->right = deserialize(str, index);
        return node;
    }
};

void deleteTree(TreeNode* root) {
    if (root) {
        deleteTree(root->left);
        deleteTree(root->right);
        delete root;
    }
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    Solution solution;
    std::string serializeroot = solution.Serialize(root);
    std::cout << serializeroot << std::endl;

    TreeNode* deserializedRoot = solution.Deserialize(serializeroot);

    deleteTree(root);
    deleteTree(deserializedRoot);

    return 0;
}

File: python/serialize.py
import sys

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def serialize(self, root: TreeNode) -> str:
        self.s = ""
        self.SerializeFunction(root)
        return self.s

    def SerializeFunction(self, root: TreeNode):
        if not root:
            self.s += "#!"
            return
        self.s += str(root.val) + "!"
        self.SerializeFunction(root.left)
        self.SerializeFunction(root.right)

    def deserialize(self, s: str) -> TreeNode:
        self.s = s
        self.index = 0
        return self.DeserializeFunction()

    def DeserializeFunction(self) -> TreeNode:
        if self.index >= len(self.s) or self.s[self.index] == "#":
            self.index += 1
            return None
        val = 0
        while self.index < len(self.s) and self.s[self.index] != "!":
            val = val * 10 + int(self.s[self.index])
            self.index += 1
        node = TreeNode(val)
        node.left = self.DeserializeFunction()
        node.right = self.DeserializeFunction()
        return node

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.right.left = TreeNode(6)
root.right.right = TreeNode(7)

solution = Solution()
s1 = solution.serialize(root)
print(s1)

deserialized_root = solution.deserialize(s1)
print(deserialized_root.val)